burn-core/src/record/serde/data.rs

```diff
@@ -0,0 +1,233 @@
+use super::de::Deserializer;
+use super::error::Error;
+use super::ser::Serializer;
+
+/// The main data structure used for deserialization.
+///
+/// It can hold tree-like structures of nested maps and vectors.
+#[derive(Debug, Clone)]
+pub enum NestedValue {
+    /// The default value, which actually does not hold any value and it is used to indicate that
+    /// the value should be populated with the default value.
+    Default,
+
+    /// A string value.
+    String(String),
+
+    /// Floating point 32-bit value.
+    F32(f32),
+
+    /// Floating point 64-bit value.
+    F64(f64),
+
+    /// Signed 16-bit integer value.
+    I16(i16),
+
+    /// Signed 32-bit integer value.
+    I32(i32),
+
+    /// Signed 64-bit integer value.
+    I64(i64),
+
+    /// Unsigned 16-bit integer value used for bf16 and f16 serialization
+    U16(u16),
+
+    /// Unsigned 64-bit integer value.
+    U64(u64),
+
+    /// A map of nested values (typically used for structs)
+    Map(HashMap<String, NestedValue>),
+
+    /// A vector of nested values (typically used for vector of structs or numbers)
+    Vec(Vec<NestedValue>),
+}
+
+impl NestedValue {
+    /// Get the nested value as a map.
+    pub fn get_map(self) -> Option<HashMap<String, NestedValue>> {
+        match self {
+            NestedValue::Map(map) => Some(map),
+            _ => None,
+        }
+    }
+
+    /// Get the nested value as a string.
+    pub fn get_string(&self) -> Option<&str> {
+        match self {
+            NestedValue::String(string) => Some(string),
+            _ => None,
+        }
+    }
+
+    /// Get the nested value as a f32.
+    pub fn get_f32(&self) -> Option<f32> {
+        match self {
+            NestedValue::F32(f32) => Some(*f32),
+            _ => None,
+        }
+    }
+
+    /// Get the nested value as a f64.
+    pub fn get_f64(&self) -> Option<f64> {
+        match self {
+            NestedValue::F64(f64) => Some(*f64),
+            _ => None,
+        }
+    }
+
+    /// Get the nested value as a i16.
+    pub fn get_i16(&self) -> Option<i16> {
+        match self {
+            NestedValue::I16(i16) => Some(*i16),
+            _ => None,
+        }
+    }
+
+    /// Get the nested value as a i32.
+    pub fn get_i32(&self) -> Option<i32> {
+        match self {
+            NestedValue::I32(i32) => Some(*i32),
+            _ => None,
+        }
+    }
+
+    /// Get the nested value as a i64.
+    pub fn get_i64(&self) -> Option<i64> {
+        match self {
+            NestedValue::I64(i64) => Some(*i64),
+            _ => None,
+        }
+    }
+
+    /// Get the nested value as a u16.
+    pub fn get_u16(&self) -> Option<u16> {
+        match self {
+            NestedValue::U16(u16) => Some(*u16),
+            _ => None,
+        }
+    }
+
+    /// Get the nested value as a u64.
+    pub fn get_u64(&self) -> Option<u64> {
+        match self {
+            NestedValue::U64(u64) => Some(*u64),
+            _ => None,
+        }
+    }
```

Comment:
as_*, e.g. `as_f32` appears more idiomatic for Enums. Here I am getting the actual inner values instead of converting values. Option usage is valid here instead of Result.